packages/ajax/src/interceptors-cache.js

```diff
@@ -0,0 +1,361 @@
+    }
+  }
+  // validate 'invalidateUrlsRegex', must be an regex expression or `falsy`
+  if (invalidateUrlsRegex) {
+    if (isGlobalOptions) {
+      throw new Error(
+        'Property `invalidateUrlsRegex` can be applied only to config per action (get, post etc)',
+      );
+    }
+    if (!(invalidateUrlsRegex instanceof RegExp)) {
+      throw new Error('Property `invalidateUrlsRegex` must be of type `RegExp` or `falsy`');
+    }
+  }
+  // validate 'requestIdentificationFn', default is url + searchParams
+  if (requestIdentificationFn) {
+    if (typeof requestIdentificationFn !== 'function') {
+      throw new Error('Property `requestIdentificationFn` must be of type `function` or `falsy`');
+    }
+  } else {
+    requestIdentificationFn = /** @param {any} data */ (
+      { url, params },
+      searchParamsSerializer,
+    ) => {
+      const serializedParams = searchParamsSerializer(params);
+      return serializedParams ? `${url}?${serializedParams}` : url;
+    };
+  }
+
+  return {
+    useCache,
+    methods,
+    timeToLive,
+    invalidateUrls,
+    invalidateUrlsRegex,
+    requestIdentificationFn,
+  };
+};
+
+/**
+ * Request interceptor to return relevant cached requests
+ * @param {function(): string} getCacheIdentifier used to invalidate cache if identifier is changed
+ * @param {GlobalCacheOptions} globalCacheOptions
+ */
+export const lionCacheRequestInterceptorFactory = (getCacheIdentifier, globalCacheOptions) => {
+  const validatedInitialLionCacheOptions = validateOptions(globalCacheOptions, true);
+
+  return /** @param {CacheRequest} cacheRequest */ cacheRequest => {
+    const { method, status, statusText, headers } = cacheRequest;
+
+    /** @type {any} */
+    let actionCacheOptions = {};
+
+    try {
+      actionCacheOptions =
+        cacheRequest.lionCacheOptions &&
+        validateOptions({
+          ...validatedInitialLionCacheOptions,
+          ...cacheRequest.lionCacheOptions,
+        });
+    } catch (e) {
+      // We need console.error here to show error somehow since Errors are not emitted from Axios
+      // eslint-disable-next-line no-console
+      console.error(e.message);
+    }
+
+    /** @type {ValidatedCacheOptions} */
+    const lionCacheOptions = {
+      ...validatedInitialLionCacheOptions,
+      ...actionCacheOptions,
+    };
+
+    // don't use cache if 'cache' === 'never'
+    if (lionCacheOptions.useCache === 'never') {
+      return cacheRequest;
+    }
+
+    const cacheId = lionCacheOptions.requestIdentificationFn(cacheRequest, searchParamSerializer);
+
+    // cacheIdentifier is used to bind the cache to the current session
+    const currentCache = getCache(getCacheIdentifier());
+    const cacheResponse = currentCache.get(cacheId, lionCacheOptions.timeToLive);
+
+    // don't use cache if the request method is not part of the configs methods
+    if (lionCacheOptions.methods.indexOf(method) === -1) {
+      // If it's NOT one of the config.methods, invalidate caches
```

Comment:
Only GET is allowed, and if the request is non GET with the same url we clean it's cache

Reply:
I got that part, but I don't get why we clean its cache? Why not ignore it, and keep the cache for consecutive calls with supported methods?

Reply:
The idea was to rather clean it immediately than keep. Support for consecutive will be added (probably) later. Some time ago we were discussing it with @daKmoR

Reply:
Let's put in the shape of a test:

```
it('returns the cached object on second call with `useCache: true`', () => {
  newCacheId();

  const indexes = addCacheInterceptors(ajax, {
    useCache: true,
  });
  const ajaxGetSpy = sinon.spy(ajax, 'get');

  return ajax
    .get('/test')
    .then(() => {
      expect(ajaxGetSpy.calledOnce).to.be.true;
      expect(ajaxGetSpy.calledWith('/test')).to.be.true;
      expect(server.requests.length).to.equal(1);
    })
    .then(() => ajax.get('/test'))
    .then(() => {
      expect(server.requests.length).to.equal(1);
    })
    .then(() => ajax.post('/test')) // ignore unsupported method
    .then(() => {
      expect(server.requests.length).to.equal(1);
    })
    .then(() => ajax.get('/test')) // return cached
    .then(() => {
      expect(server.requests.length).to.equal(1);
    })
    .finally(() => {
      ajaxGetSpy.restore();
      removeCacheInterceptors(ajax, indexes);
    });
});
```

This makes sense to me. But current behaviour would be that the last expect would equal 2, because doing unsupported method cleans the cache. I just don't see why that behaviour would be preferable. Can you explain?

Reply:
if you run a "post" on an url then it usually means the data is new so you need to fetch it... for example

```js
// foo.js
ajax.get('/test') // returns "10"

// bar.js
ajax.get('/test') // returns "10" [cached]

// foo.js later
ajax.post('/test') 

// bar.js later
ajax.get('/test') // cache is very likely no longer accurate => do request // returns "20"
ajax.get('/test') // returns "20" [cached]
```

as in, doing a post usually mutates entries in the database through the API, so it should invalidate the cache? because the cache is out of date now

alternatives for cache invalidation we looked at
- time  =>  not gonna work.... as we have time-sensitive data
- manually => is gonna be a maintenance nightmare :see_no_evil:

I would rather have cache more often invalidated than needed than having to maintain all these cache "dependencies"

Conclusion: a post request should invalidate all cache for this particular url

Reply:
Makes sense now!